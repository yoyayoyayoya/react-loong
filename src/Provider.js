import { Component, Children } from 'react'
import PropTypes from 'prop-types'
import { storeShape } from './utils/propTypes'
import warning from './utils/warning'

function warnAboutReceivingStore() {
  warning(
    `<Provider> of react-loong don't support changing the 'store' on the fly.`
  )
}

export default class Provider extends Component {
  constructor(props, context) {
    super(props, context)
    this.store = props.store
  }
  getChildContext() {
    return { store: this.store }
  }
  render() {
    return Children.only(this.props.children)
  }
}
if (process.env.NODE_ENV !== 'production') {
  Provider.prototype.componentWillReceiveProps = function(nextProps) {
    if (this.store !== nextProps.store) {
      warnAboutReceivingStore()
    }
  }
}

Provider.propTypes = {
  store: storeShape.isRequired,
  children: PropTypes.element.isRequired
}
Provider.childContextTypes = {
  store: storeShape.isRequired
}
