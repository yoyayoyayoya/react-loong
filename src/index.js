import Provider from './Provider'
import withLoong from './components/withLoong'

export { Provider, withLoong }
