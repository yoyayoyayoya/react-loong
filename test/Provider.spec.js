import React from 'react'
import renderer from 'react-test-renderer'
import PropTypes from 'prop-types'
import { Provider, withLoong } from '../src'
import { createStore } from 'loong'

/*eslint-disable no-console*/
describe('React loong', () => {
  describe('Provider', () => {
    const store = createStore({})
    @withLoong([])
    class Child extends React.Component {
      constructor(props, context) {
        expect(typeof props.publish).toEqual('function')
        super(props, context)
      }
      render() {
        return <div>test</div>
      }
    }
    Child.propTypes = {
      publish: PropTypes.func
    }
    let errorfn = null
    beforeAll(() => {
      errorfn = console.error
      console.error = m => m
    })
    afterAll(() => {
      console.error = errorfn
    })
    it('should one child only', () => {
      let error = null
      try {
        renderer.create(
          <Provider store={store}>
            <div>child</div>
            <div>child</div>
          </Provider>
        )
      } catch (e) {
        error = e
      }
      expect(error).not.toBeNull()
    })

    it('should match the store shape prop types', () => {
      let error = null
      let wrongStore = {}
      try {
        renderer.create(
          <Provider store={wrongStore}>
            <div>child</div>
            <div>child</div>
          </Provider>
        )
      } catch (e) {
        error = e
      }
      expect(error).not.toBeNull()
    })

    it('should make the child getting the publish function', () => {
      renderer.create(
        <Provider store={store}>
          <Child />
        </Provider>
      )
    })

    it('should get the warning if container change the store on the fly', () => {
      const spy = jest.spyOn(console, 'error')
      class Container extends React.Component {
        constructor(props) {
          super(props)
          this.state = { store: props.store }
          this._onClick = this._onClick.bind(this)
        }
        _onClick() {
          this.setState({ store: {} })
        }
        render() {
          return (
            <Provider store={this.state.store}>
              <a onClick={this._onClick} />
            </Provider>
          )
        }
      }
      Container.propTypes = { store: Provider.propTypes.store }

      const component = renderer.create(<Container store={store} />)
      let tree = component.toJSON()

      tree.props.onClick()
      expect(spy).toHaveBeenCalledTimes(1)
      expect(spy).toHaveBeenCalledWith(
        `<Provider> of react-loong don't support changing the 'store' on the fly.`
      )
      spy.mockReset()
      spy.mockRestore()
    })

    it('should render the component correctly', () => {
      const component = renderer.create(
        <Provider store={store}>
          <Child />
        </Provider>
      )
      let tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
