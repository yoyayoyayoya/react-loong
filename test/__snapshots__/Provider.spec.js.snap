// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React loong Provider should render the component correctly 1`] = `
<div>
  test
</div>
`;
